geomstats/learning/exponential_barycenter.py

```diff
@@ -0,0 +1,192 @@
+"""Frechet mean."""
+
+import logging
+
+from sklearn.base import BaseEstimator
+
+import geomstats.backend as gs
+from geomstats.geometry.euclidean import Euclidean
+
+EPSILON = 1e-6
+
+
+def linear_mean(points, weights=None):
+    """Compute the weighted linear mean.
+
+    The linear mean is the Frechet mean when points:
+    - lie in a Euclidean space with Euclidean metric,
+    - lie in a Minkowski space with Minkowski metric.
+
+    Parameters
+    ----------
+    points : array-like, shape=[n_samples, dimension]
+        Points to be averaged.
+
+    weights : array-like, shape=[n_samples, 1], optional
+        Weights associated to the points.
+
+    Returns
+    -------
+    mean : array-like, shape=[1, dimension]
+        Weighted linear mean of the points.
+    """
+    if isinstance(points, list):
+        points = gs.vstack(points)
+    points = gs.to_ndarray(points, to_ndim=2)
+    n_points = gs.shape(points)[0]
+
+    if isinstance(weights, list):
+        weights = gs.vstack(weights)
+    elif weights is None:
+        weights = gs.ones((n_points, ))
+
+    weighted_points = gs.einsum('...,...j->...j', weights, points)
+    mean = (gs.sum(weighted_points, axis=0) / gs.sum(weights))
+    mean = gs.to_ndarray(mean, to_ndim=2)
+    return mean
+
+
+def _default_gradient_descent(
+        group, points, weights=None, max_iter=32, step=1.,
+        epsilon=EPSILON, verbose=False):
+    """Compute the (weighted) group exponential barycenter of `points`.
+
+    Parameters
+    ----------
+    group : LieGroup
+        Instance of the class LieGroup.
+    points : array-like, shape=[n_samples, [n,n]]
+        Input points lying in the Lie Group.
+    weights : array-like, shape=[n_samples,]
+        default is 1 for each point
+        Weights of each point.
+    max_iter : int, optional (defaults to 32)
+        The maximum number of iterations to perform in the gradient descent.
+    epsilon : float, optional (defaults to 1e-6)
+        The tolerance to reach convergence. The exstrinsic norm of the
+        gradient is used as criterion.
+    step : float, optional (defaults to 1.)
+        The learning rate in the gradient descent.
+    verbose : bool
+        Level of verbosity to inform about convergence.
+
+    Returns
+    -------
+    exp_bar : array-like, shape=[n,n]
+        The exponential_barycenter of the input points.
+    """
+    n_points = points.shape[0]
+    if weights is None:
+        weights = gs.ones((n_points,))
+    sum_weights = gs.sum(weights)
+
+    def while_loop_cond(iter_index, current, grad_norm):
```

Comment:
`current` is not used, remove it?

Reply:
It was not possible to remove it with the way `while_loop` in the backend worked, but after #544 i won't use it anymore